src/adapter/src/client.rs

```diff
@@ -456,18 +456,24 @@ impl SessionClient {
 
     /// Terminates the client session.
     pub async fn terminate(&mut self) {
-        let res = self
-            .send(|tx, session| Command::Terminate {
+        // Prevent any communication with Coordinator after session is terminated.
+        // It's important to take the inner and session before sending the terminate command. If we
+        // panic and unwind after sending the Terminate command but before taking the inner and
+        // session, then we might try and send another Terminate command when we drop the client
+        // during unwinding.
+        let mut inner = self.inner.take().expect("inner invariant violated");
+        let session = self.session.take().expect("session invariant violated");
+        let res = inner
+            .send(|tx| Command::Terminate {
                 session,
                 tx: Some(tx),
             })
             .await;
-        if let Err(e) = res {
+        if let Err(e) = res.result {
             // Nothing we can do to handle a failed terminate so we just log and ignore it.
             error!("Unable to terminate session: {e:?}");
         }
-        // Prevent any communication with Coordinator after session is terminated.
-        self.inner = None;
+        self.session = Some(res.session);
```

Comment:
Can this get an explainer comment about why only the session needs to get put back? Wait, why do we need to put the session back at all? The drop impl doesn't require it to be present. Is there another reason?